### PR TITLE
Fix `IEmitCode.map`, Rewrite `MakeNDArray` in `CodeBuilder` style

### PR DESCRIPTION
**Pull request by johnc1231**
My intention is to start to replace my custom `CodePTuple` thing with the new standard `PBaseStructCode`. I've rewritten `MakeNDArray` to start doing this. 

This also fixes `IEmitCode.map`, which was broken in cases where the `CodeBuilder` ended in jumps.

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 1]
```diff
@@ -1305,10 +1306,11 @@ private class Emit[C](
...
           datat.setup,
           rowMajort.setup)
+
         val result = Code(
           shapeAddress := shapet.value[Long],
           Code.foreach(0 until nDims) { index =>
-            shapeTuple.isMissing(index).mux[Unit](
+            shapeTuple.isFieldMissing(index).mux[Unit](
               Code._fatal[Unit](s"shape missing at index $index"),
               shapeVariables(index) := shapeTuple(index))
```
Comment (johnc1231) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala:1315]: Somehow, this line is broken. If you look a few lines up, you'll see that `shapeTuple` previously referred to a `CodePTuple` but now refers to a `PBaseStructCode`. The `PBaseStructCode` has the exact same implementation of `apply` that `CodePTuple` does, and yet when I try to compile this code I get the following stack trace:

```
E           java.lang.AssertionError: assertion failed
E           	at scala.Predef$.assert(Predef.scala:156)
E           	at is.hail.asm4s.VCode.check(Code.scala:487)
E           	at is.hail.asm4s.VCode.end(Code.scala:470)
E           	at is.hail.asm4s.Code$.apply(Code.scala:53)
E           	at is.hail.asm4s.CodeLong$.$plus$extension(Code.scala:790)
E           	at is.hail.expr.types.physical.PCanonicalBaseStruct.fieldOffset(PCanonicalBaseStruct.scala:77)
E           	at is.hail.expr.types.physical.PCanonicalBaseStruct.loadField(PCanonicalBaseStruct.scala:88)
E           	at is.hail.expr.types.physical.PBaseStructCode$$anon$2.get(PBaseStruct.scala:203)
E           	at is.hail.asm4s.package$.valueToCode(package.scala:271)
E           	at is.hail.expr.ir.Emit$$anonfun$54.apply(Emit.scala:1315)
E           	at is.hail.expr.ir.Emit$$anonfun$54.apply(Emit.scala:1312)
E           	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
E           	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
E           	at scala.collection.immutable.Range.foreach(Range.scala:160)
E           	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
E           	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
E           	at is.hail.asm4s.Code$.foreach(Code.scala:128)
E           	at is.hail.expr.ir.Emit.emit(Emit.scala:1312)
E           	at is.hail.expr.ir.Emit.emit(Emit.scala:567)
E           	at is.hail.expr.ir.Emit$$anonfun$7$$anon$2.emit(Emit.scala:358)
E           	at is.hail.expr.ir.EmitUtils$$anonfun$wrapToMethod$1.apply(Emit.scala:316)
E           	at is.hail.expr.ir.EmitUtils$$anonfun$wrapToMethod$1.apply(Emit.scala:316)
E           	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
E           	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
E           	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
E           	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
.......
```

File: hail/src/main/scala/is/hail/expr/types/physical/PBaseStruct.scala [hunk 2]
```diff
@@ -193,7 +195,18 @@ abstract class PBaseStructValue extends PValue {
 abstract class PBaseStructCode extends PCode {
   def pt: PBaseStruct
 
+  val a: Code[Long]
+
+  def apply[T](i: Int): Value[T] =
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/types/physical/PBaseStruct.scala:200]: it feels a bit strange to have a constructible abstract class

[the remark this answers is not in the document]
Comment (johnc1231) [reply hail/src/main/scala/is/hail/expr/types/physical/PBaseStruct.scala:200]: Ah, that makes sense. Thanks!

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 3]
```diff
@@ -1292,61 +1292,36 @@ private class Emit[C](
...
-          cb.append(Code._println("Started making an ndarray"))
-
+        EmitCode.fromI(mb) { cb =>
           shapet.toI(cb).flatMap(cb) { case shapeTupleCode: PBaseStructCode =>
-            cb.append(Code._println("Flatmapped once"))
             datat.toI(cb).map(cb) { case dataCode: PIndexableCode =>
-              cb.append(Code._println("Mapped!"))
               val shapeTupleValue = shapeTupleCode.memoize(cb, "make_ndarray_shape")
               val dataValue = dataCode.memoize(cb, "make_ndarray_data")
               val dataPtr = dataValue.get.code.asInstanceOf[Code[Long]]
```
Comment (chrisvittal) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala:1300]: We have a helper for this.
```suggestion
              val dataPtr = dataValue.get.tcode[Long]
```

File: hail/src/main/scala/is/hail/expr/types/physical/PBaseStruct.scala [hunk 4]
```diff
@@ -193,6 +199,8 @@ abstract class PBaseStructValue extends PValue {
 abstract class PBaseStructCode extends PCode {
   def pt: PBaseStruct
 
+  val a: Code[Long]
```
Comment (chrisvittal) [inline hail/src/main/scala/is/hail/expr/types/physical/PBaseStruct.scala:202]: Unused.

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 5]
```diff
@@ -225,8 +229,12 @@ case class EmitCode(setup: Code[Unit], m: Code[Boolean], pv: PCode) {
   def toI(cb: EmitCodeBuilder): IEmitCode = {
     val Lmissing = CodeLabel()
     val Lpresent = CodeLabel()
+    cb.append(Code._println(s"PType is: ${pt.toString}"))
+    cb.append(Code._println("Pre setup"))
     cb += setup
-    cb.ifx(m, { cb.goto(Lmissing) }, { cb.goto(Lpresent) })
+    cb.append(Code._println("Post setup"))
+    cb.ifx(m, {cb.append(Code._println("Was missing")); cb.goto(Lmissing) },
+      {cb.append(Code._println("Was present")); cb.goto(Lpresent) })
```
Comment (chrisvittal) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala:237]: Remove prints

**Pull request merged**